### PR TITLE
fix: CMD+C/CTRL+C now properly copies the controller skeleton code into clipboard

### LINKED ISSUE #73: Copy Skeleton by CMD+C on Mac copys FXML into Clipboard
Originally reported by: **Alexander Casall (Bitbucket: [acasall](https://bitbucket.org/acasall), GitHub: Unknown)**

----------------------------------------

Copy Skeleton by CMD+C on Mac copys FXML into Clipboard

----------------------------------------
- Bitbucket: https://bitbucket.org/gluon-oss/scenebuilder/issue/73

### PR DESCRIPTION
**Pull request by Oliver-Loeffler**
Improved the handling of `CMD`+`c` and `CTRL`+`c` in controller skeleton preview. With the modifications, contents of the preview is properly copied into clipboard just by selecting the preview window or only the text area. Previously the `COPY` button had to be clicked.

### Issue

Fixes #73 

### Progress

- [x] Change must not contain extraneous whitespace
- [x] License header year is updated, if required
- [x] The PR name must follow the [pre-defined format](https://github.com/gluonhq/scenebuilder/blob/master/CONTRIBUTING.md)
- [x] Verify the contributor has signed [Gluon Individual Contributor License Agreement (CLA)](https://docs.google.com/forms/d/16aoFTmzs8lZTfiyrEm8YgMqMYaGQl0J8wA0VJE2LCCY)

### CODE REVIEW

File: kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonWindowController.java [hunk 1]
```diff
@@ -120,6 +126,35 @@ public SkeletonWindowController(EditorController editorController, String docume
...
+        }
+        return new KeyCharacterCombination("c", KeyCombination.CONTROL_DOWN);
+    }
+
+    private EventHandler<KeyEvent> getCopyKeyEventHandler() {
+        if (EditorPlatform.IS_MAC) {
+            return event -> handleCopyToClipboardEvent(event, "c".equals(event.getText()), event.isMetaDown());
+        } else {
+            return event -> handleCopyToClipboardEvent(event, KeyCode.C.equals(event.getCode()), event.isControlDown());
+        }
```
Comment (abhinayagarwal) [inline kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonWindowController.java:145]: use `isShortcutDown` instead

File: kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonWindowController.java [hunk 2]
```diff
@@ -130,7 +165,19 @@ public void onCloseRequest(WindowEvent event) {
     @Override
     public void openWindow() {
         super.openWindow();
-
+        /* 
+         * The event handler is only create once with the 
+         * first time the window is created. Both, the text
+         * area and the scene receive accelerators to have
+         * [CTRL]+[C] copy the contents into the clipboard.
```
Comment (abhinayagarwal) [inline kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonWindowController.java:172]: s/CTRL/SHORTCUT

File: kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonWindowController.java [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2016, 2021, Gluon and/or its affiliates.
+ * Copyright (c) 2016, 2022, Gluon and/or its affiliates.
```
Comment (abhinayagarwal) [inline kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonWindowController.java:2]: s/2022/2023

File: kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonWindowController.java [hunk 4]
```diff
@@ -120,6 +126,35 @@ public SkeletonWindowController(EditorController editorController, String docume
...
         }
+        
+        this.copyAccelerator = getCopyToClipboardKeyAccelerator();
+    }
+
+    private KeyCharacterCombination getCopyToClipboardKeyAccelerator() {
+        if (EditorPlatform.IS_MAC) {
+            return new KeyCharacterCombination("c", KeyCombination.META_DOWN);
+        }
+        return new KeyCharacterCombination("c", KeyCombination.CONTROL_DOWN);
```
[the remark this answers is not in the document]
Comment (Oliver-Loeffler) [reply kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonWindowController.java:137]: Thanks, did not know about SHORTCUT_DOWN. Will test this. If it works, this method is probably no longer needed. It was just there due to the conditional.

File: kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonWindowController.java (hunk 1 above)
Comment (Oliver-Loeffler) [reply kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonWindowController.java:145]: Same here, if it works well using `isShortcutDown`, this method can possibly ger removed. Its only there due to the conditional.

File: kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonWindowController.java [hunk 5]
```diff
@@ -120,6 +126,28 @@ public SkeletonWindowController(EditorController editorController, String docume
         if (editorController.getFxomDocument() != null) {
             editorController.getFxomDocument().sceneGraphRevisionProperty().addListener(fxomDocumentRevisionListener);
         }
+        
+        this.copyAccelerator = getCopyToClipboardKeyAccelerator();
+    }
+
+    private KeyCharacterCombination getCopyToClipboardKeyAccelerator() {
+        return new KeyCharacterCombination("c", KeyCombination.SHORTCUT_DOWN);
+    }
```
Comment (Oliver-Loeffler) [inline kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonWindowController.java:135]: @abhinayagarwal We now can even consider to remove this method at all. No need to have a method now as the conditional is gone. Still need to test on Mac. Works already fine on Windows 10.

File: kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonWindowController.java [hunk 6]
```diff
@@ -120,6 +126,28 @@ public SkeletonWindowController(EditorController editorController, String docume
...
+
+    private EventHandler<KeyEvent> getCopyKeyEventHandler() {
+        return event -> handleCopyToClipboardEvent(event, "c".equals(event.getText()), event.isShortcutDown());
+    };
```
Comment (Oliver-Loeffler) [inline kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonWindowController.java:139]: @abhinayagarwal Same here, if it works on Mac, there wont be any need to have this method. Works already fine on Windows 10. Will test on Mac soon.

File: kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonWindowController.java (hunk 5 above)
Comment (abhinayagarwal) [reply kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonWindowController.java:135]: Tested on mac. Working fine.

Comment (Oliver-Loeffler) [reply kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonWindowController.java:135]: Great to hear! My Mac is not ready yet, so seems fine to go.

File: kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonWindowController.java [hunk 7]
```diff
@@ -120,6 +126,20 @@ public SkeletonWindowController(EditorController editorController, String docume
         if (editorController.getFxomDocument() != null) {
             editorController.getFxomDocument().sceneGraphRevisionProperty().addListener(fxomDocumentRevisionListener);
         }
+        
+        this.copyAccelerator = new KeyCharacterCombination("c", KeyCombination.SHORTCUT_DOWN);
```
Comment (abhinayagarwal) [inline kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonWindowController.java:130]: Can we replace `KeyCharacterCombination` with `KeyCodeCombination` ?

Comment (Oliver-Loeffler) [reply kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonWindowController.java]: Okay, I'll replace this one. Whats the advantage of `KeyCharacterCombination` over `KeyCodeCombination`?

**Review by abhinayagarwal — APPROVED** [review]

**Comment by abhinayagarwal** [comment]
@Oliver-Loeffler Can you still run SB from `master` branch and confirm #73 still exists?

I remember copying FXML instead of the controller when I started reviewing this PR. However, I tried running from `master` again and couldn't reproduce the issue any longer.

**Comment by Oliver-Loeffler** [comment]
Yes, it was an issue mostly on Mac. Will try this one aß well tonight.

**Comment by Oliver-Loeffler** [comment]
It seems to be still there. I have this issue even on Windows 10. 
Pressing `STRG`+`C` does not do anything and one must click the `Copy` button. #73 is a thing on Windows as well. 
I've also tested it with MacOS 12.0.1 on Intel x86 and there the `CMD`+`C` is broken as well. 
So still an issue on master.

**Comment by abhinayagarwal** [comment]
Thank you for the tests. I will go ahead and merge this.

**Pull request merged**